test/nodes/test_chunker.py

```
@@ -0,0 +1,656 @@
+        assert chunker._encoder is None
+        # It should only initialize when chunk() is called
+
+    def test_start_char_incremental_tracking(self):
+        """start_char should be computed incrementally (not via O(n^2) prefix decode)."""
+        chunker = TokenChunker(chunk_size=10, chunk_overlap=0)
+
+        # Mock encoder where each character is a token and decode returns exact chars
+        call_counts = {'decode': 0}
+
+        class TrackingEncoder:
+            def encode(self, text):
```

Ruff (0.15.9)

[warning] Missing return type annotation for private function `encode`

(ANN202)